File: src/main/java/uk/co/threebugs/mochiwhattotrade3/trade/TradeRepository.java
package uk.co.threebugs.mochiwhattotrade3.trade;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.repository.query.QueryByExampleExecutor;
import org.springframework.stereotype.Repository;
import uk.co.threebugs.mochiwhattotrade3.Type;
import uk.co.threebugs.mochiwhattotrade3.account.Account;
import uk.co.threebugs.mochiwhattotrade3.setup.Setup;

import java.time.ZonedDateTime;
import java.util.List;
import java.util.Optional;

@Repository
public interface TradeRepository extends JpaRepository<Trade, Integer>, QueryByExampleExecutor<Trade> {

    List<Trade> findByTypeAndSetup_SymbolAndAccount(Type type, String symbol, Account account);

    Optional<Trade> findBySetupAndPlacedDateTimeAndAccount(Setup setup, ZonedDateTime placedDateTime, Account account);

}

File: src/main/kotlin/uk/co/threebugs/mochiwhattotrade3/trade/TradeRepository.kt
package uk.co.threebugs.mochiwhattotrade3.trade

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.repository.query.QueryByExampleExecutor
import org.springframework.stereotype.Repository
import uk.co.threebugs.mochiwhattotrade3.Type
import uk.co.threebugs.mochiwhattotrade3.account.Account
import uk.co.threebugs.mochiwhattotrade3.setup.Setup
import java.time.ZonedDateTime

@Repository
interface TradeRepository : JpaRepository<Trade, Int>, QueryByExampleExecutor<Trade> {
    fun findByTypeAndSetup_SymbolAndAccount(type: Type, symbol: String, account: Account): List<Trade>
    fun findBySetupAndPlacedDateTimeAndAccount(setup: Setup, placedDateTime: ZonedDateTime, account: Account): Trade?
}
